File: SystemCoreDrive/src/main/java/org/littletonrobotics/frc2025/subsystems/drive/ModuleIOTalonFX.java
// Copyright (c) 2025 FRC 6328
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file at
// the root directory of this project.

package org.littletonrobotics.frc2025.subsystems.drive;

import static org.littletonrobotics.frc2025.util.PhoenixUtil.tryUntilOk;

import com.ctre.phoenix6.BaseStatusSignal;
import com.ctre.phoenix6.StatusSignal;
import com.ctre.phoenix6.configs.Slot0Configs;
import com.ctre.phoenix6.configs.TalonFXConfiguration;
import com.ctre.phoenix6.controls.CoastOut;
import com.ctre.phoenix6.controls.PositionTorqueCurrentFOC;
import com.ctre.phoenix6.controls.TorqueCurrentFOC;
import com.ctre.phoenix6.controls.VelocityTorqueCurrentFOC;
import com.ctre.phoenix6.hardware.ParentDevice;
import com.ctre.phoenix6.hardware.TalonFX;
import com.ctre.phoenix6.signals.InvertedValue;
import com.ctre.phoenix6.signals.NeutralModeValue;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.util.Units;
import edu.wpi.first.units.measure.Angle;
import edu.wpi.first.units.measure.AngularVelocity;
import edu.wpi.first.units.measure.Current;
import edu.wpi.first.units.measure.Voltage;
import edu.wpi.first.wpilibj.AnalogInput;
import java.util.function.Supplier;
import org.littletonrobotics.frc2025.Constants;

public class ModuleIOTalonFX implements ModuleIO {
  private static final double driveCurrentLimitAmps = 80;
  private static final double turnCurrentLimitAmps = 40;
  public static final double driveReduction = (50.0 / 14.0) * (16.0 / 28.0) * (45.0 / 15.0);
  public static final double turnReduction = (150.0 / 7.0);

  // Hardware objects
  private final TalonFX driveTalon;
  private final TalonFX turnTalon;
  private final AnalogInput encoder;

  // Config
  private final TalonFXConfiguration driveConfig = new TalonFXConfiguration();
  private final TalonFXConfiguration turnConfig = new TalonFXConfiguration();
  private final Rotation2d encoderOffset;

  // Control requests
  private final TorqueCurrentFOC torqueCurrentRequest = new TorqueCurrentFOC(0).withUpdateFreqHz(0);
  private final PositionTorqueCurrentFOC positionTorqueCurrentRequest =
      new PositionTorqueCurrentFOC(0.0).withUpdateFreqHz(0);
  private final VelocityTorqueCurrentFOC velocityTorqueCurrentRequest =
      new VelocityTorqueCurrentFOC(0.0).withUpdateFreqHz(0);
  private final CoastOut coast = new CoastOut();

  // Inputs from drive motor
  private final StatusSignal<Angle> drivePosition;
  private final StatusSignal<AngularVelocity> driveVelocity;
  private final StatusSignal<Voltage> driveAppliedVolts;
  private final StatusSignal<Current> driveSupplyCurrentAmps;
  private final StatusSignal<Current> driveTorqueCurrentAmps;

  // Inputs from turn motor
  private final Supplier<Rotation2d> turnAbsolutePosition;
  private final StatusSignal<Angle> turnPosition;
  private final StatusSignal<AngularVelocity> turnVelocity;
  private final StatusSignal<Voltage> turnAppliedVolts;
  private final StatusSignal<Current> turnSupplyCurrentAmps;
  private final StatusSignal<Current> turnTorqueCurrentAmps;

  public ModuleIOTalonFX(DriveConstants.ModuleConfig config) {
    driveTalon = new TalonFX(config.driveMotorId(), "can_s0");
    turnTalon = new TalonFX(config.turnMotorId(), "can_s0");
    encoder = new AnalogInput(config.encoderChannel());
    encoderOffset = config.encoderOffset();
    // Configure drive motor
    driveConfig.MotorOutput.NeutralMode = NeutralModeValue.Brake;
    driveConfig.Slot0 =
        new Slot0Configs().withKP(DriveConstants.driveKp).withKI(0).withKD(DriveConstants.driveKd);
    driveConfig.Feedback.SensorToMechanismRatio = driveReduction;
    driveConfig.TorqueCurrent.PeakForwardTorqueCurrent = driveCurrentLimitAmps;
    driveConfig.TorqueCurrent.PeakReverseTorqueCurrent = -driveCurrentLimitAmps;
    driveConfig.CurrentLimits.StatorCurrentLimit = driveCurrentLimitAmps;
    driveConfig.CurrentLimits.StatorCurrentLimitEnable = true;
    driveConfig.ClosedLoopRamps.TorqueClosedLoopRampPeriod = 0.02;
    tryUntilOk(5, () -> driveTalon.getConfigurator().apply(driveConfig, 0.25));
    tryUntilOk(5, () -> driveTalon.setPosition(0.0, 0.25));

    // Configure turn motor
    turnConfig.MotorOutput.NeutralMode = NeutralModeValue.Brake;
    turnConfig.Slot0 =
        new Slot0Configs().withKP(DriveConstants.turnKp).withKI(0).withKD(DriveConstants.turnKd);
    turnConfig.Feedback.SensorToMechanismRatio = turnReduction;
    turnConfig.ClosedLoopGeneral.ContinuousWrap = true;
    turnConfig.TorqueCurrent.PeakForwardTorqueCurrent = turnCurrentLimitAmps;
    turnConfig.TorqueCurrent.PeakReverseTorqueCurrent = -turnCurrentLimitAmps;
    turnConfig.CurrentLimits.StatorCurrentLimit = turnCurrentLimitAmps;
    turnConfig.CurrentLimits.StatorCurrentLimitEnable = true;
    turnConfig.MotorOutput.Inverted =
        config.turnInverted()
            ? InvertedValue.Clockwise_Positive
            : InvertedValue.CounterClockwise_Positive;
    tryUntilOk(5, () -> turnTalon.getConfigurator().apply(turnConfig, 0.25));

    // Configure absolute encoder and set position on turn talon
    turnAbsolutePosition =
        () ->
            Rotation2d.fromRadians((double) encoder.getValue() / 3200 * 2.0 * Math.PI)
                .plus(encoderOffset);
    tryUntilOk(5, () -> turnTalon.setPosition(turnAbsolutePosition.get().getRotations(), 0.25));

    // Create drive status signals
    drivePosition = driveTalon.getPosition();
    driveVelocity = driveTalon.getVelocity();
    driveAppliedVolts = driveTalon.getMotorVoltage();
    driveSupplyCurrentAmps = driveTalon.getSupplyCurrent();
    driveTorqueCurrentAmps = driveTalon.getTorqueCurrent();

    // Create turn status signals
    turnPosition = turnTalon.getPosition();
    turnVelocity = turnTalon.getVelocity();
    turnAppliedVolts = turnTalon.getMotorVoltage();
    turnSupplyCurrentAmps = turnTalon.getSupplyCurrent();
    turnTorqueCurrentAmps = turnTalon.getTorqueCurrent();

    // Configure periodic frames
    BaseStatusSignal.setUpdateFrequencyForAll(
        1.0 / Constants.loopPeriodSecs, drivePosition, turnPosition);
    BaseStatusSignal.setUpdateFrequencyForAll(
        50.0,
        driveVelocity,
        driveAppliedVolts,
        driveSupplyCurrentAmps,
        driveTorqueCurrentAmps,
        turnVelocity,
        turnAppliedVolts,
        turnSupplyCurrentAmps,
        turnTorqueCurrentAmps);
    ParentDevice.optimizeBusUtilizationForAll(driveTalon, turnTalon);
  }

  @Override
  public void updateInputs(ModuleIO.ModuleIOInputs inputs) {
    // Update drive inputs
    inputs.driveConnected =
        BaseStatusSignal.refreshAll(
                drivePosition,
                driveVelocity,
                driveAppliedVolts,
                driveSupplyCurrentAmps,
                driveTorqueCurrentAmps)
            .isOK();
    inputs.drivePositionRad = Units.rotationsToRadians(drivePosition.getValueAsDouble());
    inputs.driveVelocityRadPerSec = Units.rotationsToRadians(driveVelocity.getValueAsDouble());
    inputs.driveAppliedVolts = driveAppliedVolts.getValueAsDouble();
    inputs.driveSupplyCurrentAmps = driveSupplyCurrentAmps.getValueAsDouble();
    inputs.driveTorqueCurrentAmps = driveTorqueCurrentAmps.getValueAsDouble();

    inputs.turnConnected =
        BaseStatusSignal.refreshAll(
                turnPosition,
                turnVelocity,
                turnAppliedVolts,
                turnSupplyCurrentAmps,
                turnTorqueCurrentAmps)
            .isOK();
    inputs.turnAbsolutePosition = turnAbsolutePosition.get().minus(encoderOffset);
    inputs.turnPosition = Rotation2d.fromRotations(turnPosition.getValueAsDouble());
    inputs.turnVelocityRadPerSec = Units.rotationsToRadians(turnVelocity.getValueAsDouble());
    inputs.turnAppliedVolts = turnAppliedVolts.getValueAsDouble();
    inputs.turnSupplyCurrentAmps = turnSupplyCurrentAmps.getValueAsDouble();
    inputs.turnTorqueCurrentAmps = turnTorqueCurrentAmps.getValueAsDouble();
  }

  @Override
  public void runDriveOpenLoop(double output) {
    driveTalon.setControl(torqueCurrentRequest.withOutput(output));
  }

  @Override
  public void runTurnOpenLoop(double output) {
    turnTalon.setControl(torqueCurrentRequest.withOutput(output));
  }

  @Override
  public void runDriveVelocity(double velocityRadPerSec, double feedforward) {
    driveTalon.setControl(
        velocityTorqueCurrentRequest
            .withVelocity(Units.radiansToRotations(velocityRadPerSec))
            .withFeedForward(feedforward));
  }

  @Override
  public void runTurnPosition(Rotation2d rotation) {
    turnTalon.setControl(positionTorqueCurrentRequest.withPosition(rotation.getRotations()));
  }

  @Override
  public void coast() {
    driveTalon.setControl(coast);
    turnTalon.setControl(coast);
  }
}

File: DriveBenchmark/src/main/cpp/drive/ModuleIOTalonFX.cpp
#include "drive/ModuleIOTalonFX.h"

#include <frc/geometry/Rotation2d.h>

#include "Robot.h"
#include "drive/DriveConstants.h"
#include "util/PhoenixUtil.h"

ModuleIOTalonFX::ModuleIOTalonFX(const int driveId, const int turnId,
                                 const int encoderId,
                                 const double encoderOffset,
                                 const bool turnInverted,
                                 const bool encoderInverted)
    : driveTalon(driveId, DriveConstants::canBus),
      turnTalon(turnId, DriveConstants::canBus),
      encoder(encoderId, DriveConstants::canBus),
      drivePosition(driveTalon.GetPosition()),
      driveVelocity(driveTalon.GetVelocity()),
      driveAppliedVolts(driveTalon.GetMotorVoltage()),
      driveSupplyCurrent(driveTalon.GetSupplyCurrent()),
      driveTorqueCurrent(driveTalon.GetTorqueCurrent()),
      turnAbsolutePosition(encoder.GetAbsolutePosition()),
      turnPosition(turnTalon.GetPosition()),
      turnVelocity(turnTalon.GetVelocity()),
      turnAppliedVolts(turnTalon.GetMotorVoltage()),
      turnSupplyCurrent(turnTalon.GetSupplyCurrent()),
      turnTorqueCurrent(turnTalon.GetTorqueCurrent()) {
  ctre::phoenix6::configs::TalonFXConfiguration driveConfig{};
  driveConfig.MotorOutput.NeutralMode =
      ctre::phoenix6::signals::NeutralModeValue::Brake;
  driveConfig.Slot0 = ctre::phoenix6::configs::Slot0Configs()
                          .WithKP(DriveConstants::driveKp)
                          .WithKD(DriveConstants::driveKd);
  driveConfig.Feedback.SensorToMechanismRatio = DriveConstants::driveReduction;
  driveConfig.TorqueCurrent.PeakForwardTorqueCurrent =
      units::ampere_t{DriveConstants::driveCurrentLimitAmps};
  driveConfig.TorqueCurrent.PeakReverseTorqueCurrent =
      -units::ampere_t{DriveConstants::driveCurrentLimitAmps};
  driveConfig.CurrentLimits.StatorCurrentLimit =
      units::ampere_t{DriveConstants::driveCurrentLimitAmps};
  driveConfig.CurrentLimits.StatorCurrentLimitEnable = true;
  driveConfig.ClosedLoopRamps.TorqueClosedLoopRampPeriod = 0.02_s;

  PhoenixUtil::tryUntilOk(
      5, [&]() { return driveTalon.GetConfigurator().Apply(driveConfig); });
  PhoenixUtil::tryUntilOk(
      5, [&]() { return driveTalon.SetPosition(0.0_rad, 0.25_s); });

  ctre::phoenix6::configs::TalonFXConfiguration turnConfig{};
  turnConfig.MotorOutput.NeutralMode =
      ctre::phoenix6::signals::NeutralModeValue::Brake;
  turnConfig.Slot0 = ctre::phoenix6::configs::Slot0Configs()
                         .WithKP(DriveConstants::turnKp)
                         .WithKD(DriveConstants::turnKd);
  turnConfig.Feedback.FeedbackRemoteSensorID = encoderId;
  turnConfig.Feedback.FeedbackSensorSource =
      ctre::phoenix6::signals::FeedbackSensorSourceValue::FusedCANcoder;
  turnConfig.Feedback.RotorToSensorRatio = DriveConstants::turnReduction;
  turnConfig.ClosedLoopGeneral.ContinuousWrap = true;
  turnConfig.TorqueCurrent.PeakForwardTorqueCurrent =
      units::ampere_t{DriveConstants::turnCurrentLimitAmps};
  turnConfig.TorqueCurrent.PeakReverseTorqueCurrent =
      -units::ampere_t{DriveConstants::turnCurrentLimitAmps};
  turnConfig.CurrentLimits.StatorCurrentLimit =
      units::ampere_t{DriveConstants::turnCurrentLimitAmps};
  turnConfig.CurrentLimits.StatorCurrentLimitEnable = true;
  turnConfig.MotorOutput.Inverted =
      turnInverted
          ? ctre::phoenix6::signals::SensorDirectionValue::Clockwise_Positive
          : ctre::phoenix6::signals::SensorDirectionValue::
                CounterClockwise_Positive;

  PhoenixUtil::tryUntilOk(5, [&]() {
    return turnTalon.GetConfigurator().Apply(turnConfig, 0.25_s);
  });

  ctre::phoenix6::configs::CANcoderConfiguration encoderConfig{};
  encoderConfig.MagnetSensor.MagnetOffset = units::radian_t{encoderOffset};
  encoderConfig.MagnetSensor.SensorDirection =
      encoderInverted
          ? ctre::phoenix6::signals::SensorDirectionValue::Clockwise_Positive
          : ctre::phoenix6::signals::SensorDirectionValue::
                CounterClockwise_Positive;

  PhoenixUtil::tryUntilOk(5, [&]() {
    return encoder.GetConfigurator().Apply(encoderConfig, 0.25_s);
  });

  ctre::phoenix6::BaseStatusSignal::SetUpdateFrequencyForAll(
      50_Hz, drivePosition, driveVelocity, driveAppliedVolts,
      driveSupplyCurrent, driveTorqueCurrent, turnAbsolutePosition,
      turnPosition, turnVelocity, turnAppliedVolts, turnSupplyCurrent,
      turnTorqueCurrent);
  PhoenixUtil::tryUntilOk(5, [&]() {
    return ctre::phoenix6::hardware::ParentDevice::OptimizeBusUtilizationForAll(
        driveTalon, turnTalon, encoder);
  });

  PhoenixUtil::registerSignals(
      true, drivePosition, driveVelocity, driveAppliedVolts, driveSupplyCurrent,
      driveTorqueCurrent, turnAbsolutePosition, turnPosition, turnVelocity,
      turnAppliedVolts, turnSupplyCurrent, turnTorqueCurrent);
};

void ModuleIOTalonFX::updateInputs(ModuleIOInputs &inputs) {
  inputs.driveConnected = ctre::phoenix6::BaseStatusSignal::IsAllGood(
      drivePosition, driveVelocity, driveAppliedVolts, driveSupplyCurrent,
      driveTorqueCurrent);
  inputs.drivePositionRad =
      drivePosition.GetValue().convert<units::radian>().value();
  inputs.driveVelocityRadPerSec =
      driveVelocity.GetValue().convert<units::rad_per_s>().value();
  inputs.driveAppliedVolts =
      driveAppliedVolts.GetValue().convert<units::volt>().value();
  inputs.driveSupplyCurrentAmps =
      driveSupplyCurrent.GetValue().convert<units::ampere>().value();
  inputs.driveTorqueCurrentAmps =
      driveTorqueCurrent.GetValue().convert<units::ampere>().value();

  inputs.turnConnected = ctre::phoenix6::BaseStatusSignal::IsAllGood(
      turnAbsolutePosition, turnPosition, turnVelocity, turnAppliedVolts,
      turnSupplyCurrent, turnTorqueCurrent);
  inputs.turnAbsolutePosition =
      frc::Rotation2d{turnAbsolutePosition.GetValue()};
  inputs.turnPosition = frc::Rotation2d{turnPosition.GetValue()};
  inputs.turnVelocityRadPerSec =
      turnVelocity.GetValue().convert<units::rad_per_s>().value();
  inputs.turnAppliedVolts =
      turnAppliedVolts.GetValue().convert<units::volt>().value();
  inputs.turnSupplyCurrentAmps =
      turnSupplyCurrent.GetValue().convert<units::ampere>().value();
  inputs.turnTorqueCurrentAmps =
      turnTorqueCurrent.GetValue().convert<units::ampere>().value();
}

void ModuleIOTalonFX::applyOutputs(const ModuleIOOutputs &outputs) {
  switch (outputs.mode) {
  case ModuleIOOutputMode::COAST:
    driveTalon.SetControl(coastRequest);
    turnTalon.SetControl(coastRequest);
    break;

  case ModuleIOOutputMode::BRAKE:
    driveTalon.SetControl(brakeRequest);
    turnTalon.SetControl(brakeRequest);
    break;

  case ModuleIOOutputMode::DRIVE:
    driveTalon.SetControl(
        velocityTorqueCurrentRequest
            .WithVelocity(
                units::radians_per_second_t{outputs.driveVelocityRadPerSec})
            .WithFeedForward(units::ampere_t{outputs.driveFeedforward}));
    if (outputs.turnNeutral) {
      turnTalon.SetControl(brakeRequest);
    } else {
      turnTalon.SetControl(positionTorqueCurrentRequest.WithPosition(
          outputs.turnRotation.Radians()));
    }
    break;

  case ModuleIOOutputMode::CHARACTERIZE:
    driveTalon.SetControl(torqueCurrentRequest.WithOutput(
        units::ampere_t{outputs.driveCharacterizationOutput}));
    turnTalon.SetControl(positionTorqueCurrentRequest.WithPosition(
        outputs.turnRotation.Radians()));
    break;
  }
}

void ModuleIOTalonFX::stop() {
  driveTalon.StopMotor();
  turnTalon.StopMotor();
}
